File: tools/scryfall-report/fetch_bulk_data.py
"""Downloads Scryfall's current oracle_cards bulk data file (one JSON object
per unique card, gzip-compressed JSONL) into ./data/. Re-run any time you
want a fresh pull - Scryfall regenerates this roughly daily.
"""

import json
import os
import urllib.request

USER_AGENT = "mtg-commander-sim-report/0.1 (personal project)"
BULK_DATA_INDEX_URL = "https://api.scryfall.com/bulk-data"
DATA_DIR = os.path.join(os.path.dirname(__file__), "data")
OUTPUT_PATH = os.path.join(DATA_DIR, "oracle-cards.jsonl.gz")


def fetch_json(url: str) -> dict:
    req = urllib.request.Request(url, headers={"User-Agent": USER_AGENT, "Accept": "application/json"})
    with urllib.request.urlopen(req) as resp:
        return json.load(resp)


def main() -> None:
    os.makedirs(DATA_DIR, exist_ok=True)

    index = fetch_json(BULK_DATA_INDEX_URL)
    oracle_entry = next(item for item in index["data"] if item["type"] == "oracle_cards")
    download_url = oracle_entry["jsonl_download_uri"]
    print(f"Downloading {oracle_entry['name']} (updated {oracle_entry['updated_at']}, "
          f"{oracle_entry['compressed_size']:,} bytes compressed)")

    req = urllib.request.Request(download_url, headers={"User-Agent": USER_AGENT})
    with urllib.request.urlopen(req) as resp, open(OUTPUT_PATH, "wb") as out:
        out.write(resp.read())

    print(f"Saved to {OUTPUT_PATH}")


if __name__ == "__main__":
    main()

File: tools/scryfall-report/fetch_bulk_data.mjs
import fs from "node:fs";
import https from "node:https";
import path from "node:path";
import { fileURLToPath } from "node:url";

const UA = "mtg-commander-sim-report/0.1 (personal project)";
const DATA_DIR = path.join(path.dirname(fileURLToPath(import.meta.url)), "data");
const OUT = path.join(DATA_DIR, "oracle-cards.jsonl.gz");
const TMP = OUT + ".tmp";

function get(url, headers = {}) {
  return new Promise((resolve, reject) => {
    https.get(url, { headers: { "User-Agent": UA, ...headers } }, (res) => {
      if (res.statusCode >= 300 && res.statusCode < 400 && res.headers.location) {
        res.resume(); return get(res.headers.location, headers).then(resolve, reject);
      }
      if (res.statusCode !== 200) { reject(new Error(`HTTP ${res.statusCode} for ${url}`)); res.resume(); return; }
      const chunks = [];
      res.on("data", (c) => chunks.push(c));
      res.on("end", () => resolve(Buffer.concat(chunks)));
    }).on("error", reject);
  });
}

const index = JSON.parse((await get("https://api.scryfall.com/bulk-data", { Accept: "application/json" })).toString("utf8"));
const oracle = index.data.find((i) => i.type === "oracle_cards");
const url = oracle.jsonl_download_uri || oracle.download_uri;
console.log(`Downloading ${oracle.name} updated ${oracle.updated_at} from ${url}`);
const body = await get(url);
console.log(`Downloaded ${body.length} bytes; gzip magic: ${body[0] === 0x1f && body[1] === 0x8b}`);
fs.writeFileSync(TMP, body);
console.log("Wrote temp:", TMP);
